File: eunji/0x08/BOJ_10799.py
import sys
from collections import deque

paren = sys.stdin.readline()
ans = 0
flag = 0
D = deque([])

for c in paren:
    if c == '(':
        ans += 1
        D.append(c)
        flag = 1
    elif (c == ')') and (flag == 1):
        ans -= 1
        if (len(D) != 1):
            ans += len(D) - 1
        flag = 0
        D.pop()
    elif (c == ')') and (flag == 0):
        D.pop()
        
print(ans)

File: eunji/0x08_stackApp/BOJ_10799.cpp
#include <iostream>
#include <deque>
using namespace std;


int main(void){
    string paren;
    cin >> paren;
    int ans, flag;
    ans = 0;
    flag = 0;
    deque<char> D;
    for (auto c : paren){
        if (c == '('){
            ans ++;
            D.push_back(c);
            flag = 1;
        }
        else if (c == ')' && flag == 1){
            ans --;
            if (D.size() != 1){        
                ans += D.size()-1;
            }
            flag = 0;
            D.pop_back();
        }
        else if (c == ')' and flag == 0){
            D.pop_back();
        }
    }
    cout << ans;
}
